incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -59,51 +58,113 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 	}
 
-	// Report "Forbidden" state and early exist if the namespace is not allowed to self-serve
-	// namespaces but has bypassed the webhook and successfully created the hns instance.
+	// Report "Forbidden" state and early exit if the namespace is not allowed to
+	// self-serve namespaces but has bypassed the webhook and successfully created
+	// the hns instance. Forbidden HNSes won't have finalizers.
 	// TODO refactor/split the EX map for 1) reconciler exclusion and 2) self-serve not allowed
 	//  purposes. See issue: https://github.com/kubernetes-sigs/multi-tenancy/issues/495
 	if EX[pnm] {
 		inst.Status.State = api.Forbidden
 		return ctrl.Result{}, r.writeInstance(ctx, log, inst)
 	}
 
-	// Get the self-serve namespace's hierarchyConfig and namespace instances.
-	nsInst, err := r.getNamespace(ctx, nm)
+	// Get the owned child namespace instance. If it doesn't exist, initialize one.
+	cnsInst, err := r.getNamespace(ctx, nm)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	switch {
-	case nsInst.Name == "":
-		log.Info("The owned subnamespace does not exist", "subnamespace", nm)
-		inst.Status.State = api.Missing
+	if deleting, err := r.onDeleting(ctx, log, inst, cnsInst); deleting {
+		return ctrl.Result{}, err
+	}
+
+	// Update the state. If the owned child namespace doesn't exist, create it.
+	// (This is how a self-service namespace is created through a CR)
+	r.updateState(log, inst, cnsInst)
+	if inst.Status.State == api.Missing {
 		if err := r.writeNamespace(ctx, log, nm, pnm); err != nil {
 			return ctrl.Result{}, err
 		}
-	case nsInst.Annotations[api.AnnotationOwner] != pnm:
-		log.Info("The owner annotation of the subnamespace doesn't match the owner", "annotation", nsInst.Annotations[api.AnnotationOwner])
+	}
+
+	// Add finalizers on all non-forbidden HNSes to ensure it's not deleted until
+	// after the owned child namespace is deleted.
+	inst.ObjectMeta.Finalizers = []string{api.MetaGroup}
+	return ctrl.Result{}, r.writeInstance(ctx, log, inst)
+}
+
+func (r *HierarchicalNamespaceReconciler) onDeleting(ctx context.Context, log logr.Logger, inst *api.HierarchicalNamespace, cnsInst *corev1.Namespace) (deleting bool, err error) {
+	// Early exit and continue reconciliation if the instance is not being deleted.
+	if inst.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	cnm := inst.Name
+	log.Info("The HNS instance is being deleted.")
+	switch {
+	case len(inst.ObjectMeta.Finalizers) == 0:
+		// We've finished process this, nothing to do.
+		log.Info("Do nothing since the finalizers are already gone.")
+		return true, nil
+	case cnsInst.DeletionTimestamp.IsZero() && r.allowsCascadingDelete(cnm):
+		// The owned namespace is not being deleted and it allows cascadingDelete.
+		// Delete the owned namespace.
+		log.Info("The subnamespace is not being deleted but it allows cascading deletion.")
+		return true, r.deleteNamespace(ctx, log, cnsInst)
+	case r.removeFinalizers(log, inst, cnsInst):
```

Comment:
This is basically a nested `switch` statement, which I find a bit hard to read. For example, if the child namespace isn't being deleted yet, I need to look in `onDeleting` to see what happens if we're allowing cascading deletion, but `removeFinalizers` to see what happens if it's not.

Combining this into one larger switch statement might make more sense here, but I'll approve this (and send it to Ryan) so it's up to you if you want to make this change or not. I'd do something like this:

```
done := false
switch{
case cnsInst.Name == "":
  log.Info("Namespace has been fully deleted")
  done = true

case cnsInst.GetAnnotations()[api.AnnotationOwner] != pnm:
  log.Info("Skipping cascading deletion: namespace has conflicting owner")
  done = true

case cnsInst.DeletionTimestamp.IsZero(): // not being deleted yet
  if r.allowsCascadingDelete(cnm) {
    if err := r.deleteNamespace(ctx, log, cnsInst); err != nil {
      return true, err
    }
  } else {
    log.Info("Skipping cascading deletion: allowsCD is false")
    done = true
  }

default: // already being deleted
  log.Info("Waiting for cascading deletion to finish")
}

if done {
  inst.ObjectMeta.Finalizers = nil
  return true, r.writeInstance(ctx, log, inst)
}

return true, nil
```

Reply:
After another thought, I decided to keep it as is to have `removeFinalizers()` separated as suggested in https://github.com/kubernetes-sigs/multi-tenancy/pull/564#discussion_r399524967.

Also there will be an additional `len(inst.ObjectMeta.Finalizers) == 0` case in the proposed large switch here, which may be too long and  hard to read.

Reply:
sgtm! Thanks.